Tests/test_function.py

```diff
@@ -287,6 +287,65 @@ def __init__(self, group=None, target=None):
 
         foo.__init__(a, target='baz')
 
+    def test_kwargs3(self):
+        """verify keyword args with keys of subtypes of str"""
+
+        class mystr(str): pass
+
+        class mymapping:
+            def __init__(self, key, val):
+                self.key = key
+                self.val = val
+            def keys(self):
+                return self.key,
+            def __len__(self):
+                return 1
+            def __getitem__(self, x):
+                return self.val
+
+        def foo(**kwargs):
+            res = type(list(kwargs.keys())[0]) == mystr and kwargs["bla"] == 2 and kwargs[mystr("bla")] == 2
+            kwargs["blip"] = 2
+            return res and kwargs["blip"] == 2
+
+        self.assertTrue(foo(**{mystr("bla"): 2}))
+        self.assertTrue(foo(**mymapping(mystr("bla"), 2)))
+
+        if is_cli:
+            from iptest.ipunittest import load_ironpython_test
+            load_ironpython_test()
+            from IronPythonTest import SplatTest1, SplatTest2
+
+            self.assertTrue(SplatTest1.FuncWithIDictKwargs(**{mystr("bla"): "bla"}))
+            self.assertTrue(SplatTest1.FuncWithIDictGenOOKwargs(**{mystr("bla"): "bla"}))
+            self.assertTrue(SplatTest1.FuncWithIDictGenSOKwargs(**{mystr("bla"): "bla"}))
+            self.assertTrue(SplatTest1.FuncWithIDictKwargs(**mymapping(mystr("bla"), "bla")))
+            self.assertTrue(SplatTest1.FuncWithIDictGenOOKwargs(**mymapping(mystr("bla"), "bla")))
+            self.assertTrue(SplatTest1.FuncWithIDictGenSOKwargs(**mymapping(mystr("bla"), "bla")))
+
+            self.assertTrue(SplatTest1.FuncWithIDictKwargs(**{"bla": "bla"}))
+            self.assertTrue(SplatTest1.FuncWithIDictGenOOKwargs(**{"bla": "bla"}))
+            self.assertTrue(SplatTest1.FuncWithIDictGenSOKwargs(**{"bla": "bla"}))
+            self.assertTrue(SplatTest1.FuncWithIDictKwargs(**mymapping("bla", "bla")))
+            self.assertTrue(SplatTest1.FuncWithIDictGenOOKwargs(**mymapping("bla", "bla")))
+            self.assertTrue(SplatTest1.FuncWithIDictGenSOKwargs(**mymapping("bla", "bla")))
+
+            # now create the sites in the reverse order of key types
+
+            self.assertTrue(SplatTest2.FuncWithIDictKwargs(**{"bla": "bla"}))
```

Comment:
Hmm, how does creating the sites in the reverse order matter? I guess my knowledge of bindings is lacking...

Reply:
It does, that's why I threw in [76453cc](https://github.com/IronLanguages/dlr/pull/242/commits/76453ccb4c3934da38c8bb04d8c2a1a1e25b8741) in the DLR. I got tripped on that.

On the first call, when the site is created, the expression to access _kwargs_ by a specific key is cached and reused on subsequent calls with the same key. But the DLR does not really fully support `Extensible<string>` as _kwarg_ keys, now simply treating them as equivalent with `string`. So without 76453cc, if the first call was with `mystr("bla")`, every subsequent call with an equivalent _kwarg_ list, say, plain `"bla"`, would still use `mystr("bla")`, which is weird and unexpected. So I let the expression cache plain `string`, which is the common denominator type for all _kwarg_ keys.

It just shows that fully supporting subclasses of `string` in the DLR is far from trivial. I am going to let `IReadOnlyDictionary<object, something>` be supported in IronLanguages/dlr#240, since this seems to be the intention, but all keys are going to arrive upcast to `string` regardless whether they were originally `string` or `Extensible<string>`.

This particular test case catches a bug that was fixed in IronLanguages/dlr#242. I wanted to have it there in case I (or somebody else) will be tempted in the future to mess around with _kwarg_ support further. And there is still more to finish.

Reply:
Might be worth including the link to the DLR PR in the comment.